tests/kernel/wave/wave_sim_test.py

```diff
@@ -210,3 +210,178 @@ def repeat(acc: tkl.Register[M, N, tkl.f32]) -> tkl.Register[M, N, tkl.f32]:
+        tkw.write(
+            a_reg,
+            c,
+            mapping=mapping,
+            elements_per_thread=STORE_ELEMS_PER_THREAD,
+        )
+
+    a = torch.randn(128, 256, dtype=torch.float32)
+    c = torch.zeros(256, 128, dtype=torch.float32)
+    transpose(a, c)
+    assert_allclose(c, a.T)
+
+
+@pytest.mark.parametrize("n", [1, 2, 4, 20])
+@pytest.mark.parametrize("c", [1, 3, 10])
+@pytest.mark.parametrize("nf", [1, 2, 8])
+@pytest.mark.parametrize("stride", [1, 2, 3])
+def test_igemm_conv(n, c, nf, stride):
+    h, w = 5, 5  # Image.
+    cf, hf, wf = c, 2, 2  # Filters.
+    x = torch.randn(n, c, h, w, dtype=torch.float32)
+    we = torch.randn(nf, cf, hf, wf, dtype=torch.float32)
+
+    padding = 0  # TODO: only pad=0 is supported for now
+    convRef = torch.nn.Conv2d(c, nf, hf, stride=stride, padding=padding, bias=False)
+    convRef.weight = torch.nn.Parameter(we)
+    out_ref = convRef(x).detach()
+
+    sym = tkl.sym
+    N, C, H, W = sym.N, sym.C, sym.H, sym.W
+    NF, HF, WF = sym.NF, sym.HF, sym.WF
+
+    H_OUT = (H + 2 * padding - HF) // stride + 1
+    W_OUT = (W + 2 * padding - WF) // stride + 1
+    SZ_OUT = H_OUT * W_OUT
+
+    x_mapping = tkw.IndexMapping(
+        lambda i, j: (
+            i // SZ_OUT,
+            j // (HF * WF),
+            (i % SZ_OUT) % W_OUT * stride + (j % (HF * WF)) % WF,
+            (i % SZ_OUT) // W_OUT * stride + (j % (HF * WF)) // WF,
+        )
+    )
+    w_mapping = tkw.IndexMapping(
+        lambda i, j: (i % NF, j // (HF * WF), j % WF, (j % (HF * WF)) // WF)
+    )
+    out_mapping = tkw.IndexMapping(
+        lambda i, j: (i // SZ_OUT, j, (i % SZ_OUT) % W_OUT, (i % SZ_OUT) // W_OUT)
+    )
+
+    K = HF * WF * C
+    M = SZ_OUT * N
+
+    # # Workgroup tile sizes
+    BLOCK_M = tkl.sym.BLOCK_M
+    BLOCK_N = tkl.sym.BLOCK_N
+    BLOCK_K = tkl.sym.BLOCK_K
+    # # Address space (for GPU, shared(1) or global(0))
+    ADDRESS_SPACE = tkl.sym.ADDRESS_SPACE
+    # Other hyperparameters
+    LOAD_ELEMS_PER_THREAD = tkl.sym.LOAD_ELEMS_PER_THREAD
+    STORE_ELEMS_PER_THREAD = tkl.sym.STORE_ELEMS_PER_THREAD
+
+    # Expose user-constraints
+    constraints: list[tkw.Constraint] = [tkw.WorkgroupConstraint(M, BLOCK_M, 0)]
+    constraints += [tkw.WorkgroupConstraint(NF, BLOCK_N, 1)]
+    constraints += [tkw.TilingConstraint(K, BLOCK_K)]
+
+    constraints += [
+        tkw.HardwareConstraint(threads_per_wave=64, waves_per_block=(1, 1, 1))
+    ]
+
+    @wave_sim(constraints)
+    def conv(
+        x: tkl.Memory[N, C, H, W, ADDRESS_SPACE, tkl.f16],
+        we: tkl.Memory[NF, C, HF, WF, ADDRESS_SPACE, tkl.f16],
+        out: tkl.Memory[N, NF, H_OUT, W_OUT, ADDRESS_SPACE, tkl.f32],
+    ):
+        c_reg = tkl.Register[M, NF, tkl.f32](0.0)
+
+        @tkw.reduction(K, init_args=[c_reg])
+        def repeat(acc: tkl.Register[M, NF, tkl.f32]) -> tkl.Register[M, NF, tkl.f32]:
+            a_reg = tkw.read(
+                x,
+                mapping=x_mapping,

```

Comment:
Something that is a little confusing here, we have x_mapping which maps a 2d index to a 4d index but the shape is only 2D. What are the semantics of mapping and shape?

Reply:
In this case, we are mapping from 4d (N, C, H, W) input to 2d (M, K) output, so mapping func takes 2 arguments (output indices) and returns 4 results (input indices). It's a good question, actually, do we want to map indices in `out->in` or `in->out` order, I choose current order because I think it will be more straightforward to handle in simulator and to translate to `vector.gather`.

Regarding question if we need an explicit shape argument, we will need to invert this mapping function to find output shape from the input shape, which should be possible in most cases, but not always.

One interesting case is manual broadcasting specific row/column:
```python
mapping = IndexMapping(lambda i, j: (i, 2))
a_reg = read(a, mapping=mapping, shape=...)
```
In this case we cannot infer resulting shape from input and mapping alone.
(We don't need this for the igemm, though)

Reply:
Okay that makes sense. I think in general indices from in -> out would make more sense, but I can see how some expressions would be easier to write from out -> in than the other way around. Okay thinking about this more, I was wondering whether you could incorporate the shape information within the IndexMapping class as a type which specifies what the index mapping does: takes indices from shape [A, B] -> [U, V, W, X] and this way we could kill two birds with one stone. You can deduce in -> out or out -> in from the types and we avoid the explicit shape argument in the read, instead keeping it inside the index mapping. What do you think?

Reply:
Updated, mapping API following offline discussion, explicit `shape` is no longer needed as it encoded directly in mapping object now.

Reply:
Awesome, thanks! Taking a look now.